### PR TITLE
Adopt WCO for Windows

### LINKED ISSUE #127449: VS Code's maximize button doesn't trigger Snap Layouts in Windows 11
Does this issue occur when all extensions are disabled?: Yes/No

- VS Code Version: 1.57.1
- OS Version: Microsoft Windows [Version 10.0.22000.51]

Steps to Reproduce:

1. Open VS Code in Windows 11 with above mentioned build.
2. Hover over the maximize button

*Snap Layouts doesn't get triggered because VS Code is using a custom title bar*

Hovering over maximize button, screenshot

### LINKED ISSUE #136132: Visual Studio code seems to have Windows 10 style window controls/decorations instead of Windows 11
I know that the window controls are not done natively because the program is made in Javascript, but the window decorations/controls are in the style of Windows 10, which only effect the maximize/minimize button.
image
VSCode controls
image
Native Windows 11 controls

### PR DESCRIPTION
**Pull request by sbatten**
This PR
fixes #136132
fixes #127449

cc @deepak1556

### CODE REVIEW

File: src/vs/platform/native/electron-main/nativeHostMainService.ts [hunk 1]
```diff
@@ -212,6 +212,18 @@ export class NativeHostMainService extends Disposable implements INativeHostMain
...
 
+	async updateTitleBarOverlay(windowId: number | undefined, backgroundColor: string, foregroundColor: string): Promise<void> {
+		const window = this.windowById(windowId);
+		if (window?.win) {
+			try {
+				window.win.setTitleBarOverlay({
+					color: backgroundColor,
+					symbolColor: foregroundColor
+				});
+			} catch { }
```
Comment (bpasero) [inline src/vs/platform/native/electron-main/nativeHostMainService.ts:223]: Not clear to me why this needs a try/catch and if so why we don't log this error?

File: src/vs/platform/windows/electron-main/window.ts [hunk 2]
```diff
@@ -246,6 +247,19 @@ export class CodeWindow extends Disposable implements ICodeWindow {
...
 				}
+
+				if (isWindows) {
+					// This logic will not perfectly guess the right colors to use on initialization,
+					// but prefer to keep things simple as it is temporary and not noticeable
+					const titleBarColor = this.themeMainService.getWindowSplash()?.colorInfo.titleBarBackground ?? this.themeMainService.getBackgroundColor();
+					const symbolColor = Color.fromHex(titleBarColor).isDarker() ? '#FFFFFF' : '#000000';
+
+					options.titleBarOverlay = {
+						height: 29,
```
Comment (bpasero) [inline src/vs/platform/windows/electron-main/window.ts:258]: Where is that `29` number coming from?

File: src/vs/workbench/electron-sandbox/parts/titlebar/titlebarPart.ts [hunk 3]
```diff
@@ -220,26 +199,8 @@ export class TitlebarPart extends BrowserTitleBarPart {
...
-			runAtThisOrScheduleAtNextAnimationFrame(() => this.adjustTitleMarginToCenter());
-
-			if (this.customMenubar) {
-				const menubarDimension = new Dimension(0, dimension.height);
-				this.customMenubar.layout(menubarDimension);
-			}
-		}
+	override updateStyles(): void {
+		super.updateStyles();
+		this.nativeHostService.updateTitleBarOverlay(this.element.style.backgroundColor, this.element.style.color);
```
Comment (bpasero) [inline src/vs/workbench/electron-sandbox/parts/titlebar/titlebarPart.ts:204]: Could we be more clever here and only go to main process when we know something changed? I am a little bit worried that `updateStyles` might be called a lot more often then needed.

File: src/vs/workbench/services/contextmenu/electron-sandbox/contextmenuService.ts [hunk 4]
```diff
@@ -97,10 +97,20 @@ class NativeContextMenuService extends Disposable implements IContextMenuService
 			let x: number;
 			let y: number;
 
-			const zoom = getZoomFactor();
+			let zoom = getZoomFactor();
 			if (dom.isHTMLElement(anchor)) {
 				const elementPosition = dom.getDomNodePagePosition(anchor);
 
+				let testElement: HTMLElement | null = anchor;
```
Comment (bpasero) [inline src/vs/workbench/services/contextmenu/electron-sandbox/contextmenuService.ts:104]: This raises a lot of questions:
- is this for all platforms or only windows?
- what does it actually do, maybe a comment would help?

**Review by bpasero — CHANGES_REQUESTED** [review]
Pretty cool 👍 

I see a bit of flicker when updating the theme, where first the WCO changes and then the workbench. Is there a way to set the WCO transparent so that we do not have to change its color and simply get the color of the title bar without having to update the WCO color?

I also see a border below the WCO when window is maximized:

image


Clicking minimize button and then restore, does not clear the button hover style for me:

image


**PS:** I did not review he CSS changes or changes in `titlebarPart.ts` because you are the expert in those by now.

File: src/vs/platform/native/electron-main/nativeHostMainService.ts (hunk 1 above)
Comment (sbatten) [reply src/vs/platform/native/electron-main/nativeHostMainService.ts:223]: When I was testing I was switching between custom and native window controls to ensure linux was good but I was still on Windows. This API will throw if there are no native controls to update, but I've added the proper checks now.

File: src/vs/platform/windows/electron-main/window.ts (hunk 2 above)
Comment (sbatten) [reply src/vs/platform/windows/electron-main/window.ts:258]: On Windows 11, we size the title bar at 30 pixels, however 1px is eaten by the native window border, so the smallest height of the title bar is 29px. I had implemented a lot of custom logic here to be completely accurate, but as above, preferring simplicity here

File: src/vs/workbench/services/contextmenu/electron-sandbox/contextmenuService.ts (hunk 4 above)
Comment (sbatten) [reply src/vs/workbench/services/contextmenu/electron-sandbox/contextmenuService.ts:104]: Added a comment to explain, it applies to all platforms. You might recall a few weeks ago that I had to fix an issue where the layout control menu was drawn in the wrong position when zoom was applied on macOS. The fix that I made is incompatible with the required css changes to support the native controls. So, this is essentially a new fix for that problem that is required with adoption of the WCO

**Comment by sbatten** [comment]
@bpasero responded to feedback and made some changes.

> Is there a way to set the WCO transparent so that we do not have to change its color and simply get the color of the title bar without having to update the WCO color?

Unfortunately no, a transparent control actually shows a native color (that pale blue from windows aero days)

> I also see a border below the WCO when window is maximized

I wasn't able to repro this one. Are you in a VM?


> Clicking minimize button and then restore, does not clear the button hover style for me

I can repro this one @deepak1556

**Review by bpasero — APPROVED** [review]

**Comment by bpasero** [comment]
Yes I am on a VM and typically have zoom level of 1 in VSCode.

**Review by raghavthind2005 — APPROVED** [review]

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 5]
```diff
@@ -67,83 +92,97 @@
...
-
-.monaco-workbench .part.titlebar > .titlebar-container.enable-title-menu > .window-title {
-	display: none;
 }
 
 /* Windows/Linux: Rules for custom title (icon, window controls)  */
-.monaco-workbench.web .part.titlebar > .titlebar-container,
-.monaco-workbench.windows .part.titlebar > .titlebar-container,
-.monaco-workbench.linux .part.titlebar > .titlebar-container {
+.monaco-workbench.web .part.titlebar>.titlebar-container,
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css:107]: Any reason you changed the style here? We typically have whitespace around `>`

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css [hunk 6]
```diff
@@ -153,90 +192,80 @@
...
+	height: 30px;
+	width: 138px;
+	zoom: calc(1 / var(--zoom-factor));
 }
 
-.monaco-workbench.mac:not(.web) .part.titlebar > .titlebar-container > .window-controls-container.show-layout-control {
-	display: flex;
+.monaco-workbench.mac .part.titlebar>.window-controls-container {
+	width: 70px;
+	height: env(titlebar-area-width, 28px);
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css:210]: I wonder why this was added?

File: src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css (hunk 5 above)
Comment (sbatten) [reply src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css:107]: some formatter I have has been doing this on occasion. and I just recently found that I had the setting which only applies this to changed regions

Comment (bpasero) [reply src/vs/workbench/browser/parts/titlebar/media/titlebarpart.css:107]: Can we revert that as part of `sbatten/outdoor-mosquito` maybe? I prefer the old style.

**Pull request merged**